mbtools/replace_im_links.py

```diff
@@ -0,0 +1,74 @@
+import json
+from mbtools import utils
+import argparse
+from lxml import html
+from pathlib import Path
+from mbtools import models
+from mbtools.fetch_im_resources import IM_PREFIX
+
+
+def parse_media_file(media_path, s3_prefix):
+    mapping = {}
+    with open(media_path) as f:
+        data = json.load(f)
+        for entry in data:
+            im_url = IM_PREFIX + entry['original_filename']
+            s3_url = s3_prefix + '/' + entry['s3_key']
+            mapping[im_url] = s3_url
+    return mapping
+
+
+def replace_im_links(content_path, media_path, s3_prefix, mode):
+    im_to_osx_mapping = parse_media_file(media_path, s3_prefix)
+    replacements = {}
+    if (mode == 'mbz'):
+        backup = utils.parse_moodle_backup(content_path)
+        activities = backup.activities()
+        q_bank = backup.q_bank
+        act_updates = 0
+        for act in activities:
+            elems = act.html_elements()
+            for elem in elems:
+                changes = utils.replace_attribute_values_tree(
+                            elem.etree_fragments[0],
+                            ['src', 'href'],
+                            im_to_osx_mapping)
+                if (len(changes) > 0):
+                    elem.update_html()
+                    act_updates += 1
+                    replacements.update(changes)
+            if act_updates > 0:
+                utils.write_etree(act.activity_filename, act.etree)
+        utils.write_etree(q_bank.questionbank_path, q_bank.etree)
+    elif (mode == 'html'):
+        for item in content_path.iterdir():
+            with open(item, 'r') as f:
+                data = f.read()
+                fragments = html.fragments_fromstring(data)
+                utils.replace_attribute_values_tree(
+                    fragments[0], ['src', 'href'], im_to_osx_mapping
+                    )
+            with open(item, 'w') as f:
```

Comment:
There's an asymmetry here that on the `mbz` code path you try to only write files that might have been modified, but on the HTML path you write out everything even if there are no changes (e.g. on one path you use the return value of `replace_attribute_values_tree` while the other you don't). This seems odd / inconsistent :man_shrugging: .